networkx/algorithms/cluster.py

```diff
@@ -42,22 +46,47 @@ def triangles(G, nodes=None):
     6
     >>> print(nx.triangles(G))
     {0: 6, 1: 6, 2: 6, 3: 6, 4: 6}
-    >>> print(list(nx.triangles(G, (0, 1)).values()))
+    >>> print(list(nx.triangles(G, [0, 1]).values()))
     [6, 6]
 
     Notes
     -----
-    When computing triangles for the entire graph each triangle is counted
-    three times, once at each node.  Self loops are ignored.
+    Self loops are ignored.
 
     """
-    # If `nodes` represents a single node in the graph, return only its number
-    # of triangles.
-    if nodes in G:
-        return next(_triangles_and_degree_iter(G, nodes))[2] // 2
-    # Otherwise, `nodes` represents an iterable of nodes, so return a
-    # dictionary mapping node to number of triangles.
-    return {v: t // 2 for v, d, t, _ in _triangles_and_degree_iter(G, nodes)}
+    if nodes is not None:
+        # If `nodes` represents a single node, return only its number of triangles
+        if nodes in G:
+            return next(_triangles_and_degree_iter(G, nodes))[2] // 2
+
+        # if `nodes` is a container of nodes, then return a
+        # dictionary mapping node to number of triangles.
+        return {v: t // 2 for v, d, t, _ in _triangles_and_degree_iter(G, nodes)}
+
+    # if nodes is None, then compute triangles for the complete graph
+
+    # dict used to avoid visiting the same nodes twice
+    # this allows calculating/counting each triangle only once
+    later_neighbors = {}
+
+    # iterate over the nodes in a graph
+    for node, neighbors in G.adjacency():
+        later_neighbors[node] = {
+            n for n in neighbors if n not in later_neighbors and n is not node
+        }
```

Comment:
you can avoid looping over this twice by using a generator:
```suggestion
    def later_neighbors_gen():
        for node, neighbors in G.adjacency():
            nbrs = {
                n for n in neighbors
                if n not in later_neighbors and n is not node
            }
            later_neighbors[node] = nbrs
            yield node, nbrs
```
and then on line 81:
```python
for node1, neighbors in later_neighbors_gen():
    ...
```

Reply:
Keeping the loop **out of a generator function** makes it easier to read and the generator doesn't save any memory.  Am I missing an advantage of `later_neighbors_gen`?  If not, I'd prefer the original to this suggestion.

Reply:
The advantage would be that the nodes are only looped over once, not twice (once on creating `later_neighbors`, once on counting triangles).

But on testing this I've just realised that the whole dictionary of `later_neighbors` is needed before the second loop can begin because it's used on line 82, so this should stay as is.